File: javascript/packages/linter/src/rules/html-no-duplicate-attributes.ts
import { ParserRule, BaseAutofixContext } from "../types.js"
import { ControlFlowTrackingVisitor, ControlFlowType, getAttributeName } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOpenTagNode, HTMLAttributeNode, ParseResult, Location } from "@herb-tools/core"

interface ControlFlowState {
  previousBranchAttributes: Set<string>
  previousControlFlowAttributes: Set<string>
}

interface BranchState {
  previousBranchAttributes: Set<string>
}

class NoDuplicateAttributesVisitor extends ControlFlowTrackingVisitor<
  BaseAutofixContext,
  ControlFlowState,
  BranchState
> {
  private tagAttributes = new Set<string>()
  private currentBranchAttributes = new Set<string>()
  private controlFlowAttributes = new Set<string>()

  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.tagAttributes = new Set()
    this.currentBranchAttributes = new Set()
    this.controlFlowAttributes = new Set()
    super.visitHTMLOpenTagNode(node)
  }

  visitHTMLAttributeNode(node: HTMLAttributeNode): void {
    this.checkAttribute(node)
  }

  protected onEnterControlFlow(_controlFlowType: ControlFlowType, wasAlreadyInControlFlow: boolean): ControlFlowState {
    const stateToRestore: ControlFlowState = {
      previousBranchAttributes: this.currentBranchAttributes,
      previousControlFlowAttributes: this.controlFlowAttributes,
    }

    this.currentBranchAttributes = new Set()

    if (!wasAlreadyInControlFlow) {
      this.controlFlowAttributes = new Set()
    }

    return stateToRestore
  }

  protected onExitControlFlow(
    controlFlowType: ControlFlowType,
    wasAlreadyInControlFlow: boolean,
    stateToRestore: ControlFlowState,
  ): void {
    if (controlFlowType === ControlFlowType.CONDITIONAL && !wasAlreadyInControlFlow) {
      this.controlFlowAttributes.forEach((attr) => this.tagAttributes.add(attr))
    }

    this.currentBranchAttributes = stateToRestore.previousBranchAttributes
    this.controlFlowAttributes = stateToRestore.previousControlFlowAttributes
  }

  protected onEnterBranch(): BranchState {
    const stateToRestore: BranchState = {
      previousBranchAttributes: this.currentBranchAttributes,
    }

    if (this.isInControlFlow) {
      this.currentBranchAttributes = new Set()
    }

    return stateToRestore
  }

  protected onExitBranch(_stateToRestore: BranchState): void {}

  private checkAttribute(attributeNode: HTMLAttributeNode): void {
    const identifier = getAttributeName(attributeNode)
    if (!identifier) return

    this.processAttributeDuplicate(identifier, attributeNode)
  }

  private processAttributeDuplicate(identifier: string, attributeNode: HTMLAttributeNode): void {
    if (!this.isInControlFlow) {
      this.handleHTMLAttribute(identifier, attributeNode)
      return
    }

    if (this.currentControlFlowType === ControlFlowType.LOOP) {
      this.handleLoopAttribute(identifier, attributeNode)
    } else {
      this.handleConditionalAttribute(identifier, attributeNode)
    }

    this.currentBranchAttributes.add(identifier)
  }

  private handleHTMLAttribute(identifier: string, attributeNode: HTMLAttributeNode): void {
    if (this.tagAttributes.has(identifier)) {
      this.addDuplicateAttributeOffense(identifier, attributeNode.name!.location)
    }

    this.tagAttributes.add(identifier)
  }

  private handleLoopAttribute(identifier: string, attributeNode: HTMLAttributeNode): void {
    if (this.currentBranchAttributes.has(identifier)) {
      this.addSameLoopIterationOffense(identifier, attributeNode.name!.location)
      return
    }

    if (this.tagAttributes.has(identifier)) {
      this.addDuplicateAttributeOffense(identifier, attributeNode.name!.location)
    }
  }

  private handleConditionalAttribute(identifier: string, attributeNode: HTMLAttributeNode): void {
    if (this.currentBranchAttributes.has(identifier)) {
      this.addSameBranchOffense(identifier, attributeNode.name!.location)
      return
    }

    if (this.tagAttributes.has(identifier)) {
      this.addDuplicateAttributeOffense(identifier, attributeNode.name!.location)
    }
  }

  private addDuplicateAttributeOffense(identifier: string, location: Location): void {
    this.addOffense(`Duplicate attribute \`${identifier}\` found on tag. Remove the duplicate occurrence.`, location)
  }

  private addSameLoopIterationOffense(identifier: string, location: Location): void {
    this.addOffense(
      `Duplicate attribute \`${identifier}\` found within the same loop iteration. Attributes must be unique within the same loop iteration.`,
      location,
    )
  }

  private addSameBranchOffense(identifier: string, location: Location): void {
    this.addOffense(
      `Duplicate attribute \`${identifier}\` found within the same control flow branch. Attributes must be unique within the same control flow branch.`,
      location,
    )
  }
}

export class HTMLNoDuplicateAttributesRule extends ParserRule {
  name = "html-no-duplicate-attributes"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoDuplicateAttributesVisitor(this.name, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_duplicate_attributes.rs
use std::collections::HashSet;

use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::control_flow_tracker::{ControlFlowTracker, ControlFlowType};
use crate::utils::tag_utils::get_attribute_name;

use herb::nodes::*;
use herb::ParseResult;
use herb::Visitor;
use herb_config::Severity;

pub struct HTMLNoDuplicateAttributesRule;

struct NoDuplicateAttributesVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  tag_attributes: HashSet<String>,
  tracker: ControlFlowTracker<HashSet<String>>,
}

impl NoDuplicateAttributesVisitor {
  fn handle_exit_control_flow(&mut self) {
    if let Some(exit_info) = self.tracker.exit_control_flow() {
      if exit_info.was_conditional {
        for attribute in &exit_info.values {
          self.tag_attributes.insert(attribute.clone());
        }
      }
    }
  }

  fn check_attribute(&mut self, attribute: &HTMLAttributeNode) {
    let identifier = match get_attribute_name(attribute) {
      Some(name) => name.to_lowercase(),
      None => return,
    };

    if !self.tracker.is_in_control_flow {
      if self.tag_attributes.contains(&identifier) {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Duplicate attribute `{}`. Browsers only use the first occurrence and ignore duplicate attributes. Remove the duplicate or merge the values.",
            identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      }

      self.tag_attributes.insert(identifier);

      return;
    }

    if self.tracker.current_control_flow_type == Some(ControlFlowType::Loop) {
      if self.tracker.current_branch_values.contains(&identifier) {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Duplicate attribute `{}` in same loop iteration. Each iteration will produce an element with duplicate attributes. Remove one or merge the values.",
            identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      } else if self.tag_attributes.contains(&identifier) {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Duplicate attribute `{}`. Browsers only use the first occurrence and ignore duplicate attributes. Remove the duplicate or merge the values.",
            identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      } else {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Attribute `{}` inside loop will appear multiple times on this element. Use a dynamic attribute name like `{}-<%= index %>` or move the attribute outside the loop.",
            identifier, identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      }
    } else {
      if self.tracker.current_branch_values.contains(&identifier) {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Duplicate attribute `{}` in same branch. This branch will produce an element with duplicate attributes. Remove one or merge the values.",
            identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      } else if self.tag_attributes.contains(&identifier) {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "Duplicate attribute `{}`. Browsers only use the first occurrence and ignore duplicate attributes. Remove the duplicate or merge the values.",
            identifier
          ),
          attribute.name.as_ref().unwrap().location.clone(),
        ));
      }

      self.tracker.control_flow_values.insert(identifier.clone());
    }

    self.tracker.current_branch_values.insert(identifier);
  }
}

impl Visitor for NoDuplicateAttributesVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    self.tag_attributes = HashSet::new();
    self.tracker = ControlFlowTracker::new();
    self.walk_html_open_tag_node(node);
  }

  fn visit_html_attribute_node(&mut self, node: &HTMLAttributeNode) {
    self.check_attribute(node);
  }

  impl_control_flow_visitor!(NoDuplicateAttributesVisitor, tracker);
}

impl Rule for HTMLNoDuplicateAttributesRule {
  fn name(&self) -> &'static str {
    "html-no-duplicate-attributes"
  }

  fn default_severity(&self) -> Severity {
    Severity::Error
  }
}

impl ParserRule for HTMLNoDuplicateAttributesRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = NoDuplicateAttributesVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      tag_attributes: HashSet::new(),
      tracker: ControlFlowTracker::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
